New rule: image accessible name descriptive (#1354)
act-rules/act-rules.github.io: _rules/image-accessible-name-descriptive-qt1vmo.md

---
id: qt1vmo
name: Image accessible name is descriptive
rule_type: atomic
description: |

>   This rule checks that images with an accessible name describe the topic or purpose.

Reviewer: I think this is grammatically incorrect.
The subject of "describe" is here "images with an accessible name" and not just "accessible name". But the rule checks that the accessible name describes the purpose of the image

Reviewer 2: English is wonderfully nuanced and easily taken the wrong way. I would suggest something more like 
```suggestion
  This rule checks that images with an accessible name describe their topic or purpose.
```

Reviewer 2: "the topic or purpose" could refer to something else on the page, not necessarily the subject of the sentence or the rule. English is wonderful for misinterpretation. "their topic or purpose" is less likely to be misinterpreted.

>   This rule checks that images with an accessible name describe their topic or purpose.

Reviewer: I still read the subject of "describe" as being "images with an accessible name" and that makes the sentence very weird (at best).

Reviewer 3: I agree with @Jym77 
My suggestion: "This rule checks that an image's topic or purpose is described by its accessible name."
Other suggestion: "This rule checks that the accessible name of an image describes the image's topic or purpose."

    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
  wcag-technique:G95: # Providing short text alternatives that provide a brief description of the non-text content
    forConformance: false
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed
input_aspects:
  - Accessibility Tree
  - DOM Tree
  - CSS Styling
  - Language
acknowledgments:
  authors:
    - Wilco Fiers
  assets:
    - W3C; HTML and W3C logo
    - Wikimedia; Fireworks image
    - Adobe; PDF logo
htmlHintIgnore:
  # https://www.npmjs.com/package/htmlhint
  # (used with `npm test` to ensure validity of code snippets)
  - 'alt-require'
  - 'attr-lowercase'
---

## Applicability

Any `img`, `canvas` or `svg` element that is [visible][] and has a non-empty [accessible name][], except if one of the following is true:

- The element has an [ancestor][] in the [flat tree][] that is [named from author][]; or
- The element is an `img` element where the [current request][]'s [state][image request state] is not [completely available][].

## Expectation

> Each test target has an [accessible name][] that serves an equivalent purpose to the [non-text content][] of that test target.

Reviewer 4: I'm a bit concerned with the difference between the expectation of this rule "... serves an equivalent purpose to ...." and the description of this rule + all the test cases examples of this rule "... describe their topic or purpose". For example, passed example 1 with the alternative text "W3C" **serve as an equivalent purpose** for the W3C logo or **describe its purpose**?
If both these statements reflect the same meaning that we should use one consistent statement in this rule.

Author: I've updated the description. I think this resolves it.

## Assumptions

> This rule assumes that the language of each test target can be correctly determined (either programmatically or by analyzing the content).

Reviewer 3: Not really sure that this is needed. The rule checks that the accessible name serves an equivalent purpose to the non-text content. How is identifying the language related?

Also, technically, you want to determine the language of the accessible name, not of the test target (which is an image, and, unless it's an image of text, won't have a language)

Reviewer: > Not really sure that this is needed. The rule checks that the accessible name serves an equivalent purpose to the non-text content. How is identifying the language related?

`<img src="bread.jpg" aria-label="pain" />` may be good or bad depending whether the language is French or English.

> This rule assumes that the language of each test target can be correctly determined (either programmatically or by analyzing the content).

Reviewer 3: ```suggestion
This rule assumes that the language of the [accessible name][] of each test target can be correctly determined (either programmatically or by analyzing the content).
```

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [Understanding Success Criterion 1.1.1: Non-text Content](https://www.w3.org/WAI/WCAG21/Understanding/non-text-content.html)
- [G94: Providing short text alternative for non-text content that serves the same purpose and presents the same information as the non-text content](https://www.w3.org/WAI/WCAG21/Techniques/general/G94)
- [G95: Providing short text alternatives that provide a brief description of the non-text content](https://www.w3.org/WAI/WCAG21/Techniques/general/G95)
- [F30: Failure of Success Criterion 1.1.1 and 1.2.1 due to using text alternatives that are not alternatives (e.g., filenames or placeholder text)](https://www.w3.org/WAI/WCAG21/Techniques/failures/F30)

## Test Cases

### Passed

#### Passed Example 1

This `img` element has an `alt` attribute that describes the image.

```html
<html lang="en">

> 	<img src="/test-assets/shared/w3c-logo.png" alt="W3C" />

Reviewer: Wondering if that should be
```suggestion
	<img src="/test-assets/shared/w3c-logo.png" alt="W3C logo" />
```
especially since most other examples seem to have an alt text of "XXX logo".

Author: I think either one passes the SC, but generally don't think it should have "logo". That's the difference between describing the image, and describing the **topic or purpose** of the image.

Reviewer: But then I don't understand why Passed Example 2 has a description that says "HTML5 logo" and not just "HTML5".

I have no problem with either, but I am a bit puzzled with the fact that in some case the description is "Foo" and in other cases it is "Foo logo". I failed to see the differences prompting this.

Not an important point at all, more curiosity on my side…

Author: Fixed the HTML 5 stuff.

#### Passed Example 2

This `svg` element has an `aria-label` attribute that describes the image.

```html
<html lang="en">
	<svg viewBox="0 0 512 512" aria-label="HTML 5 logo">
		<path
			d="M108.4 0h23v22.8h21.2V0h23v69h-23V46h-21v23h-23.2M206 23h-20.3V0h63.7v23H229v46h-23M259.5 0h24.1l14.8 24.3L313.2 0h24.1v69h-23V34.8l-16.1 24.8l-16.1-24.8v34.2h-22.6M348.7 0h23v46.2h32.6V69h-55.6"
		/>
		<path fill="#e44d26" d="M107.6 471l-33-370.4h362.8l-33 370.2L255.7 512" />
		<path fill="#f16529" d="M256 480.5V131H404.3L376 447" />
		<path
			fill="#ebebeb"
			d="M142 176.3h114v45.4h-64.2l4.2 46.5h60v45.3H154.4M156.4 336.3H202l3.2 36.3 50.8 13.6v47.4l-93.2-26"
		/>
		<path fill="#fff" d="M369.6 176.3H255.8v45.4h109.6M361.3 268.2H255.8v45.4h56l-5.3 59-50.7 13.6v47.2l93-25.8" />
	</svg>
</html>
```

#### Passed Example 3

This `canvas` element has an `aria-label` attribute that describes the image.

```html
<html lang="en">
	<canvas id="act" width="200" height="60" aria-label="ACT Rules!"></canvas>
	<script>
		const ctx = document.querySelector('#act').getContext('2d')
		ctx.font = '30px Arial'
		ctx.fillText('ACT Rules!', 20, 40)
	</script>
</html>
```

### Failed

#### Failed Example 1

> This `img` element has an `alt` attribute that incorrectly describe the image.

Reviewer 3: ```suggestion
This `img` element has an `alt` attribute that incorrectly describes the image.
```

```html
<html lang="en">
	<img src="/test-assets/shared/w3c-logo.png" alt="ERCIM" />
</html>
```

#### Failed Example 2

> This `svg` element has an `aria-label` attribute that incorrectly describe the image.

Reviewer 3: ```suggestion
This `svg` element has an `aria-label` attribute that incorrectly describes the image.
```

```html
<html lang="en">
	<svg viewBox="0 0 512 512" aria-label="W3C logo">
		<path
			d="M108.4 0h23v22.8h21.2V0h23v69h-23V46h-21v23h-23.2M206 23h-20.3V0h63.7v23H229v46h-23M259.5 0h24.1l14.8 24.3L313.2 0h24.1v69h-23V34.8l-16.1 24.8l-16.1-24.8v34.2h-22.6M348.7 0h23v46.2h32.6V69h-55.6"
		/>
		<path fill="#e44d26" d="M107.6 471l-33-370.4h362.8l-33 370.2L255.7 512" />
		<path fill="#f16529" d="M256 480.5V131H404.3L376 447" />
		<path
			fill="#ebebeb"
			d="M142 176.3h114v45.4h-64.2l4.2 46.5h60v45.3H154.4M156.4 336.3H202l3.2 36.3 50.8 13.6v47.4l-93.2-26"
		/>
		<path fill="#fff" d="M369.6 176.3H255.8v45.4h109.6M361.3 268.2H255.8v45.4h56l-5.3 59-50.7 13.6v47.2l93-25.8" />
	</svg>
</html>
```

#### Failed Example 3

> This `canvas` element has an `aria-label` attribute that incorrectly describe the image.

Reviewer 3: ```suggestion
This `canvas` element has an `aria-label` attribute that incorrectly describes the image.
```

```html
<html lang="en">
	<canvas id="act" width="200" height="60" aria-label="HTML 5 logo"></canvas>
	<script>
		const ctx = document.querySelector('#act').getContext('2d')
		ctx.font = '30px Arial'
		ctx.fillText('ACT Rules!', 20, 40)
	</script>
</html>
```

### Inapplicable

#### Inapplicable Example 1

> This `img` element has an empty (`""`) [accessible name][]. The image is described by an adjacent text.

Reviewer 3: ```suggestion
This `img` element has an empty (`""`) [accessible name][]. The image is described by the adjacent text.
```

```html
<img src="/test-assets/shared/pdf-icon.png" alt="" /> PDF document
```

#### Inapplicable Example 2

> This `img` element has an empty (`""`) [accessible name][] because of its `presentation` role.

Reviewer: Technically, "because of its `presentation` role" is incorrect in that case. Whether `role="presentation"` is here or not, the image has an empty alt text.
Then  do agree that we want to keep `role="presentation"` in the example in order to have it pass the SC…
What about
```suggestion
This decorative `img` element has an empty (`""`) [accessible name][] because it has no attributes or content to provide an accessible name.
```

```html
<html lang="en">
	<img src="/test-assets/shared/fireworks.jpg" role="presentation" />
</html>
```

#### Inapplicable Example 3

This `svg` element has an empty (`""`) [accessible name][] because it has no attributes or content to provide an accessible name.

```html
<html lang="en">
	<p>Happy new year!</p>
	<svg height="200" xmlns="http://www.w3.org/2000/svg">
		<polygon points="100,10 40,180 190,60 10,60 160,180" fill="yellow" />
	</svg>
</html>
```

#### Inapplicable Example 4

This `canvas` element has an empty (`""`) [accessible name][] because it has no attributes or content to provide an accessible name.

```html
<html lang="en">
	<p>Happy new year!</p>
	<canvas id="newyear" width="200" height="200"></canvas>
	<script>
		const ctx = document.querySelector('#newyear').getContext('2d')
		ctx.fillStyle = 'yellow'
		ctx.beginPath()
		ctx.moveTo(100, 10)
		ctx.lineTo(40, 180)
		ctx.lineTo(190, 60)
		ctx.lineTo(10, 60)
		ctx.lineTo(160, 180)
		ctx.fill()
	</script>
</html>
```

#### Inapplicable Example 5

This `img` element is not [visible][].

```html
<html lang="en">

> 	<img src="/test-assets/shared/w3c-logo.png" alt="W3C" style="display:none" />

Reviewer: ```suggestion
	<img src="/test-assets/shared/w3c-logo.png" alt="W3C logo" style="display:none" />
```
(only if the same change is made on Passed Example 1: keep them in sync)

</html>
```

#### Inapplicable Example 6

This `canvas` element is not [visible][] because it is completely transparent.

```html
<html lang="en">
	<canvas width="200" height="200"></canvas>
</html>
```

#### Inapplicable Example 7

> This `img` element is has no [accessible name][] because it is not [included in the accessibility tree][].

Reviewer 3: ```suggestion
This `img` element has no [accessible name][] because it is not [included in the accessibility tree][].
```

```html
<html lang="en">

> 	<img aria-hidden="true" src="/test-assets/shared/fireworks.jpg" alt="some picgure" />

Reviewer 3: ```suggestion
	<img aria-hidden="true" src="/test-assets/shared/fireworks.jpg" alt="some picture" />
```

Not that it makes any difference

Reviewer: Actually, I'd still go for a descriptive name in that case. "some picture" is one of the less descriptive alt text possible…
```suggestion
	<img aria-hidden="true" src="/test-assets/shared/fireworks.jpg" alt="fireworks" />
```

</html>
```

#### Inapplicable Example 8

This `svg` element is ignored because it is a child of a link that provides its [accessible name][].

```html
<a href="//w3.org" aria-label="W3C Website">
	<svg height="200" xmlns="http://www.w3.org/2000/svg" aria-label="star">
		<polygon points="100,10 40,180 190,60 10,60 160,180" fill="yellow" />
	</svg>

> </a>

Reviewer 3: I know it is inapplicable, but still, AT users will be able to access the `svg` accessible name and they will get "star" which is totally unrelated to the link. We could use the image of the W3C logo in this example so this does not happen.

Author: That's not an accessibility issue though "star" correctly describes the SVG here. The link target is ambiguous to everyone.

```

#### Inapplicable Example 9

> This `img` element has an `src` attribute which will cause the [image request state][] to be [Broken](https://html.spec.whatwg.org/#img-error).

Reviewer 3: ```suggestion
This `img` element has a `src` attribute which will cause the [image request state][] to be [Broken](https://html.spec.whatwg.org/#img-error).
```
